hivemind/client/averaging/__init__.py

```diff
@@ -279,6 +289,134 @@ async def rpc_aggregate_part(self, stream: AsyncIterator[averaging_pb2.Averaging
+                    yield averaging_pb2.DownloadData(tensor_part=part, metadata=metadata)
+                    metadata = None
+                else:
+                    yield averaging_pb2.DownloadData(tensor_part=part)
+
+    def get_current_state(self) -> Tuple[Any, Sequence[torch.Tensor]]:
+        """
+        Get current state and send it to a peer. executed in the host process. Meant to be overriden.
+        :returns: a tuple of (serializable_small_metadata, sequence of torch tensors)
+        """
+        with self.get_tensors() as tensors:
+            return dict(group_key=self.get_group_bits()), tensors
+
+    async def _get_current_state_from_host_process(self):
+        """ Executed in the averager process inside rpc_download_state """
+        future, _future = MPFuture.make_pair()
+        self._pipe.send(('_TRIGGER_GET_CURRENT_STATE', _future))
+        return await future
+
+    def _background_thread_fetch_current_state_if_asked(self):
+        """ Executed in the host process as a background thread. """
+        while True:
+            trigger, future = self.pipe.recv()
+            assert trigger == '_TRIGGER_GET_CURRENT_STATE'
+            try:
+                state_metadata, state_tensors = self.get_current_state()
+                # note: serialize here to avoid initializing cuda in the guest process
+                state_metadata = pickle.dumps(state_metadata)
+                state_tensors = tuple(tensor.cpu().detach().requires_grad_(tensor.requires_grad)
+                                      for tensor in state_tensors)
+                future.set_result((state_metadata, state_tensors))
+            except BaseException as e:
+                future.set_exception(e)
+                logger.warning(e)
+                continue
+
+    def load_state_from_peers(self, wait=True) -> Optional[Any]:
+        """ Try to download the latest optimizer state one of the existing peer """
+        future, _future = MPFuture.make_pair()
+        self.pipe.send(('_load_state_from_peers', [], dict(future=_future)))
+        return future.result() if wait else future
+
+    async def _load_state_from_peers(self, future: MPFuture):
+        key_manager = self._matchmaking.group_key_manager
+        peers_info, _ = self.dht.get(f"{key_manager.prefix}.all_averagers", latest=True) or ({}, None)
+        metadata = None
+        if not isinstance(peers_info, dict):
+            logger.info("Averager could not load state from peers: peer dict is corrupted.")
+            future.set_result(None)
+            return
+
+        def get_priority(peer):
+            info = peers_info[peer]
+            if not isinstance(info, ValueWithExpiration) or not isinstance(info.value, (float, int)):
+                logger.warning(f"Skipping averager {peer} - bad peer info {info.value} (expected a number).")
+                return -float('inf')
+            return float(info.value)  # prefer peers with latest update time
```

Comment:
Can we rewrite this function not to use external variables and to move it outside the method? For example, with functools.partial, if you intend to use it as a key. It's harder to reason about closures than about non-nested functions, and in this case it's not that necessary